swiftnav/src/reference_frame/mod.rs

```diff
@@ -284,108 +400,187 @@ impl fmt::Display for TransformationNotFound {
+            transformations: TransformationGraph::new(),
         }
-        TransformationGraph { graph }
     }
 
-    /// Get the shortest path between two reference frames, if one exists
+    /// Create a repository from a list of transformations
     ///
-    /// This function will also search for reverse paths if no direct path is found.
-    /// The search is performed breadth-first.
+    /// # Note
+    ///
+    /// If there are duplicated transformations in the list, the
+    /// last one in the list will take priority
+    pub fn from_transformations<T: IntoIterator<Item = Transformation>>(
+        transformations: T,
+    ) -> Self {
+        let mut repo = Self::new();
+        repo.extend(transformations);
+        repo
+    }
+
+    /// Create a repository with the builtin transformations
     #[must_use]
-    pub fn get_shortest_path(
+    pub fn from_builtin() -> Self {
+        Self::from_transformations(builtin_transformations())
+    }
+
+    /// Add a transformation to the repository
+    ///
+    /// This will rebuild the internal graph to include the new transformation.
+    pub fn add_transformation(&mut self, transformation: Transformation) {
+        let from = transformation.from;
+        let to = transformation.to;
+        let params = transformation.params;
+        let inverted_params = params.invert();
+
+        self.transformations
+            .entry(from.clone())
+            .or_default()
+            .extend([(to.clone(), params)]);
+
+        // Add inverted parameters as well
+        self.transformations
+            .entry(to)
+            .or_default()
+            .extend([(from, inverted_params)]);
+    }
+
+    /// Transform a [`Coordinate`] to a new reference frame
+    ///
+    /// This function finds the shortest series of transformations from the coordinate's
+    /// initial reference frame to the requestest one, then sequentially applies
+    /// those transformations to get the new position and velocity. The epoch of the
+    /// coordinate is not modified in this process.
+    ///
+    /// # Errors
+    ///
+    /// [`TransformationNotFound`] is returned as an error if no path from the
+    /// coordinate's reference frame to the requested reference frame could be found
+    /// in the repository.
+    pub fn transform(
+        &self,
+        coord: &Coordinate,
+        to: &ReferenceFrame,
+    ) -> Result<Coordinate, TransformationNotFound> {
+        let epoch = coord.epoch().to_fractional_year_hardcoded();
+
+        let apply_transformation = |(position, velocity): (ECEF, Option<ECEF>),
+                                    params: &TimeDependentHelmertParams|
+         -> (ECEF, Option<ECEF>) {
+            params.transform(&position, velocity.as_ref(), epoch)
+        };
+
+        let (position, velocity) = self
+            .get_shortest_path(coord.reference_frame(), to)?
+            .into_iter()
+            .fold((coord.position(), coord.velocity()), apply_transformation);
```

Comment:
HUGE NIT: I personally find the assigning of the closure a bit confusing to read and prefer to let the compiler infer the types here. If you disagree please just disregard here 😅 


```suggestion
        let (position, velocity) = self
            .get_shortest_path(coord.reference_frame(), to)?
            .into_iter()
            .fold(
                (coord.position(), coord.velocity()),
                |(position, velocity), params| {
                    params.transform(&position, velocity.as_ref(), epoch)
                },
            );
```